Add Stepper Indicator Component

jorytindall opened the pull request:
### :pushpin: Summary

Initial component generation and variant/state styling for the stepper indicator.

### :hammer_and_wrench: Detailed description

Configured the initial stepper indicator component with `status` types, `states`, `isInteractive` boolean, and `isProcessing` state with loading Flight Icon. Additionally I added examples of the different statuses and variants when paired together in the showcase.

### :camera_flash: Screenshots
<img width="585" alt="Screen Shot 2022-07-20 at 9 48 25 AM" src="https://user-images.githubusercontent.com/2200899/180038532-cee27751-4b64-47c7-b7c5-48a6868dd7c2.png">


### :link: External links

<!-- Issues, RFC, etc. -->

***

### 👀 How to review

<!-- Suggest how it's best for the reviewer to review the code (choose one, or remove) -->
👉 Review commit-by-commit
👉 Review by files changed

Reviewer's checklist:

- [ ] +1 Percy if applicable
- [ ] Confirm that PR has a changelog update via [Changesets](https://github.com/changesets/changesets) if needed

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/design-system?pullRequest=470)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/design-system?pullRequest=470)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/design-system?pullRequest=470).</sub>

MelSumner commented on packages/components/addon/components/hds/stepper/stepper-indicator/index.js:
```diff
@@ -0,0 +1,103 @@
+// export const DEFAULT_TYPE = 'step'
+// export const DEFAULT_NUMBER = '1'
+
+export const STATUSES = ["incomplete", "inProgress", "complete"];
```
Tip: if you run `yarn lint`, you'll see these issues before you push. 

There are some other options available as well (see package.json for the full list):

- `yarn lint:hbs` this will lint handlebars (`.hbs`) files
- `yarn lint:hbs --fix` this will lint `.hbs` files and also auto fix anything that is auto-fixable
- `yarn lint:js` this will lint `.js` files
- `yarn lint:js --fix` this will lint and also fix any issues that are auto-fixable

Don't worry if you forget, I often have a second commit that is "fixing linter" related 👍

MelSumner commented on packages/components/app/styles/components/stepper/stepper-indicator.scss:
```diff
@@ -0,0 +1,169 @@
+        }
+    }
+
+    &:not(.isInteractive) {
```
comment: I _think_ we were previously trying to avoid using `:not`; @didoo can confirm though.

MelSumner commented on packages/components/addon/components/hds/stepper/stepper-indicator/index.hbs:
```diff
@@ -0,0 +1,22 @@
+<div class={{this.classNames}} ...attributes>
```
issue (a11y, blocker): This will need some sort of role, depending on the intent of the design.

MelSumner commented on packages/components/addon/components/hds/stepper/stepper-indicator/index.hbs:
```diff
@@ -0,0 +1,22 @@
+<div class={{this.classNames}} ...attributes>
+  <div class="svg-container">
+    <svg width="20" height="23" viewBox="0 0 20 23" fill="none" xmlns="http://www.w3.org/2000/svg">
```
question: is this for the background of the step indicator?

MelSumner commented on packages/components/addon/components/hds/stepper/stepper-indicator/index.hbs:
```diff
@@ -0,0 +1,22 @@
+    </svg>
+  </div>
+  <div class="indicator-status">
+    {{#if @isProcessing}}
+        <FlightIcon @name="loading" @size="16" />
+      {{else if (eq @status "complete")}}
+        <FlightIcon @name="check" @size="16" />
+      {{else if (eq @status "incomplete")}}
+        <span>{{@stepNumber}}</span>
+      {{else if (eq @status "inProgress")}}
+        <span>{{@stepNumber}}</span>
+    {{/if}}
```
question: is `status` only checked if `isProcessing` is true? If so, this section needs a little bit of refactoring; use a nested `{{#if}}` here, and put the default state into the `{{else}}` at the end before the closing `{{/if}}`. Here's some documentation: https://guides.emberjs.com/release/components/conditional-content/#toc_usage

MelSumner commented:
I wonder if this PR should be converted to a draft? It is still missing some essential things: 

- documentation
- tests
- linting

There's a checklist here for more clarity: https://github.com/hashicorp/design-system/blob/main/packages/components/NEW-COMPONENT-CHECKLIST.md

didoo commented on packages/components/tests/dummy/app/router.js:
```diff
@@ -36,6 +36,10 @@ Router.map(function () {
     this.route('toast');
+
+    this.route('stepper', function () {
+      this.route('stepper-indicator');
```
probably we are going to collect `stepper-item` and `stepper-indicator` under the same page; this will require a small refactoring but we can do it later

jorytindall replied:
Yes, I updated the routing to the following:

* Stepper (reserving for if/when we build a larger stepper pattern)
  * Indicator
    * Step
    * Task

Does that work for hierarchy? So the component usage looks like `Hds::Stepper::Indicator::Step`

didoo reviewed (commented):
hey @jorytindall I've added a few comments, to start :) but I think is better if the next step is a pair-programming session together so we can go through the comments and other simplifications in the code that can be done

jorytindall commented on packages/components/addon/components/hds/stepper/stepper-indicator/index.js:
```diff
@@ -0,0 +1,103 @@
+    let { type = "step" } = this.args;
+
+    return type;
+  }
```
Removed this and moved the type into it's own component.

jorytindall commented:
@didoo I made a bunch of changes and added an additional `task` component, here's a summary of my changes:
* Cleaned up and optimized the styling and naming conventions based on your feedback and comments
* Added a `Indicator::Task` component to break apart the Step and Task components
* Restructured the folder structure and naming conventions to better align with how the indicator will be consumed and how it could be used in a stepper item component
    * `Hds::Stepper::Indicator::Step` and `Hds::Stepper::Indicator::Task`. This adds a few more folders, but I think it’s more explicit, previously the naming convention was `Hds::Stepper::StepperIndicator::Step` which is repetitive.
* Restructured the /stepper/indicator showcase on the scrappy site to render both the Step and the Task component with their relevant status properties and interactive states. 
    * Added .mock-styles to use in the showcase 
* Updated some of the generated tests which pass on my local instance but fail with I push my repo 🤔

didoo reviewed (commented):
@jorytindall I had another go, nothing major for me. seeing more and more improvements in the implementation. 👍

jorytindall commented on packages/components/addon/components/hds/stepper/indicator/step.hbs:
```diff
@@ -0,0 +1,19 @@
+  </div>
+  <div class="hds-stepper-indicator-step__status">
+    {{#if (eq @status "processing")}}
+      <FlightIcon class="hds-stepper-indicator-step__icon" @name="loading" @size="16" />
```
We decided the `loading` icon is more explicit and simpler for this use case, especially since the icon is smaller.

jorytindall commented on packages/components/addon/components/hds/stepper/indicator/step.js:
```diff
@@ -0,0 +1,54 @@
+   * @default false
+   */
+
+  isInteractive = this.args.isInteractive;
```
Ok got it, thanks for the context and reasoning.

jorytindall commented on packages/components/app/styles/components/stepper/indicator/task.scss:
```diff
@@ -0,0 +1,65 @@
+
+@each $status in $hds-stepper-indicator-task-statuses {
+    // Interactive variant
+    .hds-stepper-indicator-task--is-interactive {
```
Got it, ok that makes sense

jorytindall commented on packages/components/app/styles/components/stepper/indicator/task.scss:
```diff
@@ -0,0 +1,65 @@
+    height: 12px;
+}
+
+@each $status in $hds-stepper-indicator-task-statuses {
```
Yes, I think I can clean up the step a bit to use the same convention. Adding the props block should reduce `status` and `state` repitition.

jorytindall commented on packages/components/tests/dummy/app/router.js:
```diff
@@ -36,6 +36,9 @@ Router.map(function () {
       this.route('standalone');
     });
     this.route('toast');
+    this.route('stepper', function () {
```
Yep, I can do that. To me it made sense to maintain the `/indicator` in the path since that's what we currently have, but I'll reorganize it.

didoo replied:
We have something similar for the `Card` component. Technically the component is `Hds::Card::Container`, but in anticipation of a more complex `Card` component, we have set the path to `card`: https://design-system-components-hashicorp.vercel.app/components/card

jorytindall commented on packages/components/tests/dummy/app/routes/components/stepper/indicator.js:
```diff
@@ -0,0 +1,10 @@
+import Route from '@ember/routing/route';
+
+import { STATUSES } from '@hashicorp/design-system-components/components/hds/stepper/indicator/step';
```
Ok, I kept the naming convention in the `js` file the same between `step` and `task`, but updated the import in the route model to reflect the specific component.

didoo replied:
I don't understand :) 

(I don't see any change... 🤔)

jorytindall commented on packages/components/tests/dummy/app/templates/components/stepper/indicator.hbs:
```diff
@@ -0,0 +1,111 @@
+</section>
+
+<section>
+  <h3 class="dummy-h3" id="component-api"><a href="#component-api" class="dummy-link-section">§</a> Component API</h3>
```
I've started to add some of the content, unless you think this is getting a little long/lots of comments, might make sense to keep the same PR?

didoo replied:
yes, I see now the content is added, so all good

jorytindall commented on packages/components/tests/integration/components/hds/stepper/indicator/step-test.js:
```diff
@@ -0,0 +1,26 @@
+      assert
+        .dom('#test-stepper-indicator-step')
+        .hasClass('hds-stepper-indicator-step');
+    });
```
As covered in our 1:1, I'll try writing the testing first and reach out if I have any questions or get stuck.

didoo commented on packages/components/addon/components/hds/stepper/indicator/step.hbs:
```diff
@@ -0,0 +1,19 @@
+<div class={{this.classNames}} ...attributes role="listitem" aria-label={{@status}}>
+  <div class="hds-stepper-indicator-step__svg-container" aria-hidden="true">
```
+1 on what @alex-ju suggests

didoo replied:
also, what if we call this class `hds-stepper-indicator-step__svg-hexagon`? so it's even clearer what it is

didoo commented on packages/components/addon/components/hds/stepper/indicator/step.js:
```diff
@@ -0,0 +1,55 @@
+import { assert } from '@ember/debug';
+
+export const DEFAULT_STATUS = 'incomplete';
+
```
extra space
```suggestion
```

didoo commented on packages/components/addon/components/hds/stepper/indicator/task.hbs:
```diff
@@ -0,0 +1,11 @@
+<div class={{this.classNames}} ...attributes role="listitem" aria-label={{@status}}>
+  {{#if (eq @status "incomplete")}}
+    <FlightIcon class="hds-stepper-indicator-task__icon" @name="circle" @size="16" />
+  {{else if (eq @status "progress")}}
+    <FlightIcon class="hds-stepper-indicator-task__icon" @name="circle-half" @size="16" />
+  {{else if (eq @status "processing")}}
+    <FlightIcon class="hds-stepper-indicator-task__icon" @name="loading" @size="16" />
+  {{else if (eq @status "complete")}}
+    <FlightIcon class="hds-stepper-indicator-task__icon" @name="check-circle" @size="16" />
+  {{/if}}
```
[non-blocker/suggestion] since the only difference between all these conditions is the icon name, what if we compute its value in the backing JavaScript class, similar to what is done with the `className`?

This would lead to this code:
```
<div class={{this.classNames}} ...attributes role="listitem" aria-label={{@status}}>
  <FlightIcon class="hds-stepper-indicator-task__icon" @name=={{this.iconName}} @size="16" />
</div>
```
which is much easier to read, and reduce the risk of bugs.

didoo commented on packages/components/addon/components/hds/stepper/indicator/task.js:
```diff
@@ -0,0 +1,55 @@
+import { assert } from '@ember/debug';
+
+export const DEFAULT_STATUS = 'incomplete';
+
```
extra space
```suggestion
```

didoo commented on packages/components/app/styles/components/stepper/indicator/step.scss:
```diff
@@ -0,0 +1,184 @@
+@each $status in $hds-stepper-indicator-step-statuses {
+  // For each status of the non-interactive variant, set the text color, svg fill, and svg stroke colors based on $non-interactive-props
+  .hds-stepper-indicator-step--status-#{$status} {
+    .hds-stepper-indicator-step__text,
+    .hds-stepper-indicator-step__icon {
```
[non-blocker/suggestion] I _think_ you can replace these two selectors with a single one: `hds-stepper-indicator-step__status` (also below in the file)

jorytindall replied:
Yep, worked fine!

didoo commented on packages/components/app/styles/components/stepper/indicator/task.scss:
```diff
@@ -0,0 +1,63 @@
+}
+
+.hds-stepper-indicator-task__icon {
+    color: inherit;
```
this is not needed
```suggestion
```

didoo commented on packages/components/app/styles/components/stepper/indicator/task.scss:
```diff
@@ -0,0 +1,63 @@
+
+// Base styling for indicator::task
+.hds-stepper-indicator-task {
+    width: $hds-stepper-indicator-task-size;
+    height: $hds-stepper-indicator-task-size;
+    position: relative;
+    display: flex;
+    justify-content: center;
+    align-items: center;
+    color: var(--token-color-foreground-strong);
```
please sort props alphabetically (I know it's a pain; we will soon introduce linting for SCSS and this should become automatic).

didoo commented on packages/components/app/styles/components/stepper/indicator/task.scss:
```diff
@@ -0,0 +1,63 @@
+    @each $status in $hds-stepper-indicator-task-statuses {
+        // For each status set the icon color based on the $status-props
+        &.hds-stepper-indicator-task--status-#{$status} {
+            color: map-get($status-props, $status, "color-default");
+            &:hover, &.mock-hover { color: map-get($status-props, $status, "color-hover")};
+            &:active, &.mock-active { color: map-get($status-props, $status, "color-active")};
```
better to move on multiple lines, it's more readable (it would have been easier to spot that a couple of semicolons are in the wrong place)

```suggestion
            color: map-get($status-props, $status, "color-default");

            &:hover,
            &.mock-hover {
                color: map-get($status-props, $status, "color-hover");
            }

            &:active,
            &.mock-active {
                color: map-get($status-props, $status, "color-active");
            }
```

didoo commented on packages/components/tests/dummy/app/routes/components/stepper.js:
```diff
@@ -0,0 +1,11 @@
+import Route from '@ember/routing/route';
+
+import { STATUSES as STEP_STATUSES } from '@hashicorp/design-system-components/components/hds/stepper/indicator/step';
+import { STATUSES as TASK_STATUSES } from '@hashicorp/design-system-components/components/hds/stepper/indicator/task';
```
nice!

didoo commented:
> Wrote some tests 🧪! Tried to not duplicate myself, let me know if there's some coverage that I missed or if I should reformat, etc.

@jorytindall I _think_ you didn't push them though :)

didoo reviewed (commented):
@jorytindall added a bunch of comments, mostly nitpicking. I just need to review the updated integration tests, but overall the PR is ready to be merged for me. nice job Jory, really impressed 🙌

didoo commented on packages/components/addon/components/hds/stepper/indicator/task.hbs:
```diff
@@ -1,11 +1,7 @@
+  <FlightIcon class="hds-stepper-indicator-task__icon" @name={{this.iconName}} @size="16" />
+</div>
+
+{{!-- <div class={{this.classNames}} ...attributes role="listitem" aria-label={{@status}}>
```
@jorytindall extra commented code here?

jorytindall replied:
Whoops, yep, sorry about that

didoo commented on packages/components/addon/components/hds/stepper/indicator/task.js:
```diff
@@ -35,6 +34,29 @@ export default class HdsStepperIndicatorTaskIndexComponent extends Component {
+   */
+
+  get iconName() {
+    let status = this.args.status;
```
[issue] Done in this way you are missing the default value (`DEFAULT_STATUS`) because you're reading the argument. If you instead you use `this.status` is already taken in account for you by the component class itself (see code example below).

didoo commented on packages/components/addon/components/hds/stepper/indicator/task.js:
```diff
@@ -35,6 +34,29 @@ export default class HdsStepperIndicatorTaskIndexComponent extends Component {
+
+    // Check the status using a ternary operator and returning a icon name value
+    const value =
+      status === 'incomplete'
+        ? 'circle'
+        : status === 'progress'
+        ? 'circle-half'
+        : status === 'processing'
+        ? 'loading'
+        : status === 'complete'
+        ? 'check-circle'
+        : null;
```
maybe you could use a mapping object at the top, and then return the value given the key? see for example how it's done here: https://github.com/hashicorp/design-system/blob/580c8ddf765fbd3fff03f42868c1f696f72df9cd/packages/components/addon/components/hds/alert/index.js#L9-L15

I mean something like this:

```
const MAPPING_STATUS_TO_ICON = {
  'incomplete': 'circle',
  'progress': 'circle-half',
  'processing': 'loading',
  'complete': 'check-circle',
};

...

get iconName() {
  return MAPPING_STATUS_TO_ICON(this.status);
}      
```

jorytindall replied:
Great, exactly what I was looking for. I was wondering whether a map would work better rather than a cascading ternary statement.

didoo replied:
> whether a map would work better rather than a cascading ternary statement.

Are you _really_ asking this? 🤔

(just kidding 🤣)

didoo reviewed (commented):
@jorytindall a couple of points to check

didoo commented on packages/components/addon/components/hds/stepper/indicator/task.js:
```diff
@@ -3,6 +3,13 @@ import { assert } from '@ember/debug';
 
 export const DEFAULT_STATUS = 'incomplete';
 export const STATUSES = ['incomplete', 'progress', 'processing', 'complete'];
+export const ICONS = ['circle', 'circle-half', 'loading', 'check-circle'];
```
pretty sure this line is not needed
```suggestion
```

didoo reviewed (approved):
Good to go for me (only one line to remove).

🛳 it!

PS: for the Percy errors, you have to rebase your branch on `main`, or merge `main` in your branch, which one you're used to.

PPS: **important**: don't forget to add a changeset (see instructions [here](https://github.com/hashicorp/design-system/blob/main/packages/components/NEW-COMPONENT-CHECKLIST.md#component-review:~:text=add%20a%20changelog%20update%20via%20Changesets%20if%20needed%20using%20the%20command%20yarn%20changeset%20(in%20the%20project%20root)): add a changelog update via [Changesets](https://github.com/changesets/changesets) if needed using the command `yarn changeset` (in the project root))

MelSumner commented:
@jorytindall great work on this component!  It looks like it could have used a final a11y review- can you DM me so we can collab on getting these fixes in before 1.0? 

cc/ @didoo @Dhaulagiri
